### PR TITLE
make the decode / decrypt pipeline non-allocating (groundwork)

### PR DESCRIPTION
**Pull request by japaric**
this PR lays the groundwork for making the decoding - decryption - decoding pipeline "non-allocating" AKA removing as many allocations as possible

at a very high level that is enabled by these changes

- `MessageDeframer::pop` performs decoding - decryption in-place and yields borrowed data (slices) to the next layer
- `Codec` gains a lifetime parameter `'a` to allow its `read` method to return maximally borrowed data (`&'a`) from `Reader<'a>` -- this lets the _post_-decryption decoding to also be non-allocating
- `Payload` becomes a `std::Cow`-like enum to support carrying either borrowed or owned data -- `HandshakePayload` and other types indirectly become `Cow`-like as well
- `State::handle` is tweaked to allow its return type hold borrowed data contained in the `message` parameter

to showcase how this enables the removal of allocations from the pipeline, this PR makes decoding & decryption of application-data as well as certificates (`CertificateChain`) non-allocating

removing the rest of the allocations that exist in the pipeline will be done in a subsequent PR

breaking changes:
- this change the signature of `MessageDecrypter::decrypt`. actual implementations don't have to change much, as you can see in the diff, but a breaking change is still a breaking change.

this PR builds on top of #1595 

--- 

I think it's worthwhile to explain why `Payload` and many of the `State` implementers need to become `Cow`-like and cannot contain just references.

In very simplified terms, you can think of the process of advancing `Connection`'s internal state machine as the following function: (this applies to both the buffered `Connection` and `UnbufferedConnection` APIs)

```rust
struct Connection {
    state: Option<Box<dyn State +'static>>,
    deframer: MessageDeframer,
    // ..
}

// in impl Connection
fn process_tls_records<'a>(&mut self, buffer: &mut DeframerSliceBuffer<'a>) {
    let mut current = self.state.take().unwrap();

    while let Some(message) = deframer.pop(buffer);{
        // message has type BorrowedPlainMessage<'a>
        let next = current.handle(state, message);
        current = next;
    }

    self.state = Some(current.into_owned());
}
```

each time this function is called, _several_ records are decoded and decrypted from the buffered incoming TLS data (the `buffer` parameter in the above snippet). as mentioned above, `Deframer::pop` now yields slices with lifetime `'a`. each `message`s that's fed into the state machine through the `State::handle` method can change the concrete state type behind the `current` trait object; each of those intermediate states can hold the `'a` borrowed data in `message` thanks to the `State::handle` change done in this PR.

eventually, the `deframer` is exhausted (e.g. the user may need to read data from the socket and append it to `buffer`) and it's time to return from the function. however, just after exiting the `while` loop `current` may contain borrowed `'a` content. it's not possible to store that borrowed content in `Connection` because its `state` field has a `'static` constraint. that's where we are forced to erase the `'a` lifetime by switching the true state behind `current` into its owned variant; this heap allocates any borrowed content. it's worth noting that `into_owned` it's a no-operation in the case the state type does not contain any borrowed data.

so yes, there's still one potential allocation in this function call but all of the many decoding and decryption events that happen within this function call becoming non-allocating; it also becomes possible to move between states -- within a single function call -- without needing to allocate borrowed content.

### CODE REVIEW

**Comment by japaric** [comment]
@cpu  @ctz @djc could I get a quick sanity check on the direction here? the main commits that affect the end-user and/or internal _interfaces_ are:

- make Reader work on mutable slices
- make MessageDecrypter::decrypt operate on slices (:warning: this is a semver breaking change)
- make Deframer::pop return a slice (+)
- prepare Codec for non-allocating decoding
- add a Borrowed variant to Payload
- prepare State trait to handle slice fields

TL;DR make things `std::Cow`-like and add a bunch of lifetimes everywhere

the bulk of commits that will reduce allocations will be fairly mechanical like the 'reduce allocs around ServerCertDetails ' one

thoroughly minimizing the internal API in an incremental fashion is going to take a lot of commits but I'll break out a smaller PR out of this one that only does the interface changes

now, I mainly did this to enable in-place (non-allocating) decoding / decryption in the `UnbufferedConnection` API but this PR also improves the performance of the existing `Connection` API.

so far, I have only removed allocations from -- what my uneducated guess of what the biggest contributors are like -- application-data and certificates and I'm seeing improvements of this magnitude (ran these locally; I've filtered out the noisy `aws_lc_rs` benchmarks)

| Scenario                                          | Baseline | Candidate |                Diff |
| ------------------------------------------------- | -------: | --------: | ------------------: |
| handshake_tickets_ring_1.2_rsa_aes_client         |  4615881 |   4548840 |  ✅ -67041 (-1.45%) |
| handshake_session_id_ring_1.2_rsa_aes_server      |  4327962 |   4267674 |  ✅ -60288 (-1.39%) |
| handshake_tickets_ring_1.2_rsa_aes_server         |  4776473 |   4715679 |  ✅ -60794 (-1.27%) |
| handshake_session_id_ring_1.3_ecdsa_chacha_client | 41598210 |  41489715 | ✅ -108495 (-0.26%) |
| handshake_tickets_ring_1.3_ecdsa_chacha_client    | 41796305 |  41688683 | ✅ -107622 (-0.26%) |
| handshake_tickets_ring_1.3_rsa_chacha_client      | 42040674 |  41932666 | ✅ -108008 (-0.26%) |
| handshake_tickets_ring_1.3_rsa_aes_client         | 42083707 |  41975702 | ✅ -108005 (-0.26%) |
| handshake_tickets_ring_1.3_ecdsa_aes_client       | 41837011 |  41731569 | ✅ -105442 (-0.25%) |
| handshake_session_id_ring_1.3_ecdsa_aes_client    | 41651062 |  41546453 | ✅ -104609 (-0.25%) |
| handshake_session_id_ring_1.3_rsa_aes_client      | 41899822 |  41798231 | ✅ -101591 (-0.24%) |
| handshake_session_id_ring_1.3_rsa_chacha_client   | 41845117 |  41743759 | ✅ -101358 (-0.24%) |

---

(+) I intend to use this `RawSlice` type in `Context` to make the `UnbufferedConnection` truly not-allocating when it comes to decrypting app-data and it'll probably also become instrumental to allowing caller-side certificate verification with the `UnbufferedConnection` API for future compatibility with async cert verifiers

**Comment by cpu** [comment]
> @cpu @ctz @djc could I get a quick sanity check on the direction here?

ACK'ing this but wanted to mention it might be a little bit. We're focused on 0.22 release prep and Ctz is AFK for a little bit. The high-level vision sounds good to me, but I think input from Djc and Ctz will be most valuable here. My experience in this domain is more limited.

**Comment by japaric** [comment]
> but I'll break out a smaller PR out of this one that only does the interface changes

I, instead, trimmed down this PR and it's ready for review. I have added a high level description of the changes in the PR description. removing the bulk of the allocations will be need in a separate / subsequent PR.

(the MSRV seems to not like the function/struct update syntax used in this PR :cry: )

File: rustls/src/conn.rs [hunk 1]
```diff
@@ -649,31 +651,40 @@ impl<Data> ConnectionCore<Data> {
...
         };
 
-        let mut borrowed_buffer = deframer_buffer.borrow();
-        while let Some(msg) = self.deframe(Some(&*state), &mut borrowed_buffer)? {
+        let mut discard = 0;
+        loop {
+            let mut borrowed_buffer = deframer_buffer.borrow();
+            borrowed_buffer.queue_discard(discard);
+
+            let opt_msg = self.deframe(Some(&*state), &mut borrowed_buffer)?;
```
Comment (ctz) [inline rustls/src/conn.rs:659]: I think this continues to have the same issue as the other PR? `borrowed_buffer.pending_discard()` is crucial data that must not be lost on error.

**Comment by ctz** [comment]
@rustls-benchmarking bench

File: rustls/src/msgs/message.rs [hunk 2]
```diff
@@ -387,6 +387,7 @@ impl TryFrom<PlainMessage> for Message {
 ///
 /// This type also cannot decode its internals and
 /// cannot be read/encoded; only `OpaqueMessage` can do that.
+#[derive(Debug)]
 pub struct BorrowedPlainMessage<'a> {
```
Comment (cpu) [inline rustls/src/msgs/message.rs:429]: Let's roll the `Debug` derive into b99e527499e96ada274bf4a29383b88b8b27fa22

File: rustls/src/msgs/deframer.rs [hunk 3]
```diff
@@ -155,19 +155,21 @@ impl MessageDeframer {
 
             // If it's not a handshake message, just return it -- no joining necessary.
             if msg.typ != ContentType::Handshake {
+                let message = msg.into_owned();
```
Comment (cpu) [inline rustls/src/msgs/deframer.rs]: I'm not sure I spot how these `MessageDeframer` changes map to the `Decrypted` update. Possibly a candidate to move into a different commit?

File: rustls/src/msgs/message.rs [hunk 4]
```diff
@@ -211,19 +228,6 @@ impl<'a> BorrowedOpaqueMessage<'a> {
...
-    pub(crate) fn into_owned(self) -> OpaqueMessage {
-        let Self {
-            typ,
-            version,
-            payload,
-        } = self;
-        OpaqueMessage {
-            typ,
-            version,
-            payload: Payload::new(&*payload),
```
Comment (cpu) [inline rustls/src/msgs/message.rs]: It looks like while this was moved the `&*` was dropped, can you fix that in the commit that added `into_owned` on `BorrowedOpaqueMessage` instead?

File: rustls/src/msgs/handshake.rs [hunk 5]
```diff
@@ -1352,12 +1363,17 @@ impl Codec<'_> for CertificateEntry {
...
+            cert: certificate_der_into_owned(CertificateDer::read(r)?),
             exts: Vec::read(r)?,
         })
     }
 }
 
+// FIXME(perf) sub-optimal in the owned case; switch to `CertificateDer::into_owned` when available
+fn certificate_der_into_owned(cert: CertificateDer<'_>) -> CertificateDer<'static> {
+    CertificateDer::from(cert.to_vec())
+}
```
Comment (cpu) [inline rustls/src/msgs/handshake.rs:1375]: I think this FIXME can be resolved, I see [CertificateDer::into_owned](https://docs.rs/rustls-pki-types/latest/rustls_pki_types/struct.CertificateDer.html#method.into_owned) landed.

**Review by cpu — COMMENTED** [review]
This is really cool.

File: rustls/src/msgs/deframer.rs (hunk 3 above)
Comment (pvdrz) [reply rustls/src/msgs/deframer.rs]: That's because this `msg` value comes from a `Decrypted` value, which has a `BorrowedPlainMessage` now instead of a `PlainMessage`

File: rustls/src/conn.rs (hunk 1 above)
Comment (pvdrz) [reply rustls/src/conn.rs:659]: maybe here we can just match on the output of deframe and discard on error. What do you think?

Comment (ctz) [reply rustls/src/conn.rs:659]: I think it would be more robush if `deframer_buffer.borrow()` returned an object that automatically took care of this on drop. However, I think achieving that is not currently possible because the deframer buffer is mutable (despite deframing being a strictly read-only operation) and fixing that is a larger refactor.

So I guess ensuring every caller of `deframer_buffer.borrow()` does the right thing is the second best option now.

**Comment by ctz** [comment]
> to showcase how this enables the removal of allocations from the pipeline, this PR makes decoding & decryption of application-data as well as certificates (`CertificateChain`) non-allocating

I think we're just moving the allocations around rather than eliminating them?

eg, for `CertificateChain`, in TLS1.3 client:

- the deframer message decoding is now allocation-free,
- in `ExpectCertificate`: from `Message` -> `ServerCertDetails` is allocation-free,
- (if this is the end of the flight, the contents of `ServerCertDetails` are copied, but that is infrequent -- however we do need test coverage in both cases)
- in `ExpectCertificateVerify`, we copy (both shallow and deep if needed) the `CertificateChain` into `cx.common.peer_certificates` to appear in the public API.

Total copies of the certificate data: 1, plus one extra small vec

Prior to this PR:

- the deframer message decoding allocates for the `CertificateChain` contents
- in `ExpectCertificate`: this is moved from `Message` into `ServerCertDetails` which is allocation-free,
- in `ExpectCertificateVerify` the `CertificateChain` is moved into `cx.common.peer_certificates` to appear in the public API.

Total copies of the certificate data: 1

File: rustls/src/conn.rs (hunk 1 above)
Comment (pvdrz) [reply rustls/src/conn.rs]: Ok I did the second best thing :P

File: rustls/src/msgs/message.rs (hunk 2 above)
Comment (ctz) [reply rustls/src/msgs/message.rs:432]: yes, usually a "fixup" commit is for squashing into a previous commit

Comment (cpu) [reply rustls/src/msgs/message.rs:432]: I don't think it was described as a fixup prefixed commit when I left this feedback, but if it was I apologize for missing it!

Comment (pvdrz) [reply rustls/src/msgs/message.rs:432]: I moved this hunk to the commit you mentioned. Is that what you wanted?

File: rustls/src/msgs/message.rs [hunk 6]
```diff
@@ -368,16 +381,38 @@ impl Message {
...
+        let Self { version, payload } = self;
+        Message {
+            version,
+            payload: payload.into_owned(),
+        }
+    }
+}
+
+// NOTE does not avoid allocations like the `BorrowedPlainMessage -> Message` conversion. this is
+// only used in (integration) tests
```
Comment (ctz) [inline rustls/src/msgs/message.rs:395]: Seems like `Message::try_from(PlainMessage)` is used in `first_handshake_message`, which is not just an integration test?

That seems OK (avoiding it seems to require quite a bit of change on the `Acceptor` API side) but this comment is potentially misleading.

**Review by ctz — COMMENTED** [review]
A few more comments, mainly about commit structuring and clarity. But I think this is getting closer.

File: rustls/src/msgs/deframer.rs [hunk 7]
```diff
@@ -438,22 +439,26 @@ pub struct DeframerSliceBuffer<'a> {
     // a fully initialized buffer that will be deframed
     buf: &'a mut [u8],
     // number of bytes to discard from the front of `buf` at a later time
-    discard: usize,
```
[the remark this answers is not in the document]
Comment (pvdrz) [reply rustls/src/msgs/deframer.rs:441]: I think I'm not following you. The `discard` field type is changed to `Cell<usize>` on this commit basically. this is done to avoid borrowing issues later but I think adding that to a later PR would make it a little bit confusing. I'll add an explicit `groundwork:` to it.

**Review by ctz — APPROVED** [review]

**Review by cpu — APPROVED** [review]
Thanks! All of my earlier feedback looks resolved and I didn't have anything new to add.

File: rustls/src/msgs/codec.rs [hunk 8]
```diff
@@ -85,6 +85,68 @@ impl<'a> Reader<'a> {
...
+
+    fn take(&mut self, length: usize) -> Option<&'a mut [u8]> {
+        if self.left() < length {
+            return None;
+        }
+        let (taken, rest) = mem::take(&mut self.buffer).split_at_mut(length);
+        self.used += taken.len();
+        self.buffer = rest;
+        Some(taken)
+    }
```
Comment (djc) [inline rustls/src/msgs/codec.rs:146]: Nit: I feel like these should remain in the same order as for `Reader`?

File: rustls/src/msgs/message.rs [hunk 9]
```diff
@@ -214,6 +190,66 @@ impl OpaqueMessage {
...
+
+        Ok(Self {
+            typ,
+            version,
+            payload,
+        })
+    }
+}
+
+fn read_opaque_message_minus_payload(
```
Comment (djc) [inline rustls/src/msgs/message.rs:252]: Naming nit: I'd probably call this `read_opaque_message_header()`.

File: rustls/src/msgs/deframer.rs [hunk 10]
```diff
@@ -491,6 +491,15 @@ fn copy_into_buffer(buf: &mut [u8], src: &[u8], at: usize) {
 }
 
 trait FilledDeframerBuffer {
+    fn filled_get_mut<I>(&mut self, index: I) -> &mut I::Output
```
Comment (djc) [inline rustls/src/msgs/deframer.rs:602]: Style nit: `impl SliceIndex<[u8]>`?

File: rustls/src/msgs/base.rs [hunk 11]
```diff
@@ -57,6 +57,36 @@ impl<'a> DerefMut for BorrowedPayload<'a> {
...
+    pub(crate) fn pop(&mut self) -> Option<u8> {
+        if self.is_empty() {
+            None
+        } else {
+            let len = self.len();
+            let last = self[len - 1];
+            self.truncate(len - 1);
+            Some(last)
+        }
+    }
```
Comment (djc) [inline rustls/src/msgs/base.rs]: Style nit: this is a pretty surprising method order IMO.

File: rustls/src/msgs/base.rs [hunk 12]
```diff
@@ -57,6 +57,36 @@ impl<'a> DerefMut for BorrowedPayload<'a> {
...
+    pub(crate) fn new(bytes: &'a mut [u8]) -> Self {
+        Self(bytes)
+    }
+
+    pub(crate) fn into_inner(self) -> &'a mut [u8] {
+        self.0
+    }
+
+    pub(crate) fn pop(&mut self) -> Option<u8> {
+        if self.is_empty() {
```
Comment (djc) [inline rustls/src/msgs/base.rs:103]: Should do an early return here?

File: rustls/src/record_layer.rs [hunk 13]
```diff
@@ -62,12 +62,12 @@ impl RecordLayer {
     /// an error is returned.
     pub(crate) fn decrypt_incoming(
         &mut self,
-        encr: OpaqueMessage,
+        encr: BorrowedOpaqueMessage,
```
Comment (djc) [inline rustls/src/record_layer.rs]: Nit: let's make the lifetime explicit?

File: rustls/src/msgs/deframer.rs [hunk 14]
```diff
@@ -462,15 +491,28 @@ impl<'a> DeframerSliceBuffer<'a> {
     pub fn is_empty(&self) -> bool {
         self.len() == 0
     }
+
+    fn take(&mut self, raw: RawSlice) -> &'a mut [u8] {
+        let start = (raw.ptr as usize)
+            .checked_sub(self.buf.as_ptr() as usize)
+            .unwrap();
```
Comment (djc) [inline rustls/src/msgs/deframer.rs:543]: So we're doing pointer arithmetic here -- how do we guarantee that `raw.ptr` and `self.buf` are from the same allocation? Even though there is no `unsafe` code here, this seems to be potentially brittle code which should come with more documentation.

File: rustls/src/msgs/deframer.rs [hunk 15]
```diff
@@ -504,6 +549,13 @@ impl<'a> DeframerSliceBuffer<'a> {
 
         &mut taken[start..]
     }
+
+    fn raw_slice_to_filled_range(&self, raw: RawSlice) -> Range<usize> {
+        let adjust = self.discard - self.taken;
+        let start = ((raw.ptr as usize).checked_sub(self.buf.as_ptr() as usize)).unwrap() - adjust;
```
Comment (djc) [inline rustls/src/msgs/deframer.rs:555]: Same here.

**Review by djc — COMMENTED** [review]
Hmm, I'm not really comfortable with some of the things that are happening here. In particular the work around the `RawSlice` which has some invariants which are neither enforced by the type system nor clearly documented, and some complexity around the `AppendPayload` trait and impls that it's not obvious are needed.

(Informed by my experience working on #1179.)

File: rustls/src/msgs/deframer.rs (hunk 10 above)
Comment (cpu) [reply rustls/src/msgs/deframer.rs:602]: I think because of the `Output` associated type being referred to in the `return` using `impl` ends up being pretty ugly. The return needs to be qualified to avoid ambiguity:

```rust
fn filled_get_mut(
        &mut self,
        index: impl SliceIndex<[u8], Output = [u8]>,
    ) -> &mut <RangeFrom<usize> as SliceIndex<[u8]>>::Output 
```

I'm inclined to leave it as-is with the generic parameter. WDYT?

File: rustls/src/msgs/codec.rs (hunk 8 above)
Comment (cpu) [reply rustls/src/msgs/codec.rs]: Fixed.

File: rustls/src/msgs/message.rs (hunk 9 above)
Comment (cpu) [reply rustls/src/msgs/message.rs]: Renamed

File: rustls/src/msgs/base.rs (hunk 11 above)
Comment (cpu) [reply rustls/src/msgs/base.rs]: Reordered

File: rustls/src/record_layer.rs (hunk 13 above)
Comment (cpu) [reply rustls/src/record_layer.rs]: Done. A later commit further refines this to a named lifetime.

File: rustls/src/msgs/base.rs (hunk 12 above)
Comment (cpu) [reply rustls/src/msgs/base.rs:107]: Fixed.

File: rustls/src/msgs/deframer.rs (hunk 10 above)
Comment (djc) [reply rustls/src/msgs/deframer.rs:602]: Yeah, the impl argument doesn't make sense, sorry. I'd probably inline the I bound in the declaration instead of having a where clause, too.

Comment (cpu) [reply rustls/src/msgs/deframer.rs]: Fixed.

File: rustls/src/msgs/deframer.rs (hunk 14 above)
Comment (cpu) [reply rustls/src/msgs/deframer.rs:543]: I'm struggling to understand the `RawSlice` abstraction and the commit that introduced it is missing any motivating context or guidance in the commit message (I guess that's part of what you're flagging in this review...). I gather we want to use pointers here instead of a normal slice into `self.buf` for reasons related to the lifetime of the slice and the `DeframerSliceBuffer` buffer, but I can't exactly put my finger on it.

>  how do we guarantee that raw.ptr and self.buf are from the same allocation?

I haven't been able to think of a way to tie the `RawSlice` back to the buffer it came from that won't trip into the lifetime issues that I think this design was built around avoiding. Do you have any suggestions?

>  this seems to be potentially brittle code which should come with more documentation.

Would additional documentation about the invariants be enough to satisfy your concerns? That's certainly doable :+1:

Comment (cpu) [reply rustls/src/msgs/deframer.rs:543]: > how do we guarantee that raw.ptr and self.buf are from the same allocation?

We could drop the `RawSlice::from<&[u8]>` constructor and implement a method on `DeframerSliceBuffer` that takes a slice argument and asserts the base ptr and len of the slice falls within the `DeframerSliceBuffer`'s `buf` before giving back the `RawSlice`.

Writing that out, I'm not sure that it's a very useful change w.r.t safety or brittleness since it just changes where we panic: constructing the `RawSlice` or doing the `split_at_mut`/`checked_sub` in `take`/`raw_slice_to_filled_range` with the result of the pointer arithmetic.

@ctz Do you have any thoughts on this feedback or suggestions for refactoring?

Comment (ctz) [reply rustls/src/msgs/deframer.rs:543]: I'd say we're doing a similar thing in a different way elsewhere inside msgs/deframer.rs: but with `Range` rather than pointer arithmetic.

That is, at least, more obvious -- though I think it still suffers from similar fragility problems if anyone changes the buffer while the `Range` is valid. That instance is especially egregious because the buffer is under control of the application while the `Range` is valid. I guess this is one of the negative results of pursuing a 100% (including handshaking) no-allocation solution here rather than just addressing the main data path.

Comment (ctz) [reply rustls/src/msgs/deframer.rs:543]: I have an extremely broken and half-baked branch with some refactoring in this area, but haven't reached an ending on it (either achieving something good, or giving up!)

Some of the ideas I've so far failed to achieve:

- The deframer only ever reads from its buffer, and the `Deframed` object it returns doesn't keep a borrow of the message data; it just tells you where it is. That gives the deframer potential to become an iterator over `Result<Deframed, Error>` items in the future, maybe.
- Someone with a `Deframed` message can decrypt it in-place by providing the `record_layer` and the mutable slice containing it. That process should also not keep a mutable borrow of the buffer.
- The handshake part of the deframer is split out into a separate thing. At the moment the deframer is doing both slicing up TLS messages, and also gluing together TLS handshake protocol fragments. I would hope this smooths out the complexity hotspot in the deframer.
- The QUIC code can directly use the handshake-part-of-the-deframer, which should relieve us of a lot of the `if quic { }` decisions in the deframer.

Comment (djc) [reply rustls/src/msgs/deframer.rs:543]: That sounds great. Maybe @cpu could take the half-baked branch and run with it?

Comment (ctz) [reply rustls/src/msgs/deframer.rs:543]: Pushed https://github.com/rustls/rustls/tree/jbp-refactor-deframer

Comment (cpu) [reply rustls/src/msgs/deframer.rs:543]: > That sounds great. Maybe @cpu could take the half-baked branch and run with it?

Sure, I can give that a go :+1:, but is it a blocker for landing this work? I'd rather not have to maintain two unfinished branches at the same time if possible. If it's OK with you folks my preference would be to land this branch and then rebase the refactor on top to continue.

> Pushed https://github.com/rustls/rustls/tree/jbp-refactor-deframer

@ctz Could you maybe summarize where you left off to help me get started?

Comment (cpu) [reply rustls/src/msgs/deframer.rs:543]: > That seems reasonable.

Thanks :+1: I'll add some additional documentation around the `RawSlice` invariants and then merge this.

Comment (cpu) [reply rustls/src/msgs/deframer.rs:548]: Updated with additional doc comments.

File: rustls/src/msgs/deframer.rs (hunk 15 above)
Comment (cpu) [reply rustls/src/msgs/deframer.rs:564]: Updated with additional doc comments.

**Comment by cpu** [comment]
> Hmm, I'm not really comfortable with some of the things that are happening here

Pulling some discussion [out of a resolved review thread](https://github.com/rustls/rustls/pull/1597#discussion_r1446139734):

* I've added more documentation around the invariants required by the `RawSlice` portions of the work internal to the deframer.
* There's similar pre-existing "brittleness" w.r.t the deframer and `Range` operations.
* We're going to land this branch as-is, accepting that it introduces a bit more brittleness in the short term.
* I'm going to pick up [a refactor branch](https://github.com/rustls/rustls/tree/jbp-refactor-deframer) that ctz started to try and reign in some of the complexity as follow-up work.
